dotconfig/i3/config.org
#+title: Config

* About this config
This is a fork of my sway config, changed for i3.
- [-] TODO [20%]
  - [ ] Clipboard manager
  - [-] XanTools [1/4]
    - [X] XanShoot
    - [X] XanBackgrounds
    - [ ] XanSwayMove
    - [ ] XanSwaySwitch
  - [ ] Password manager
  - [ ] Polybar and Polybar Toggle
  - [X] Rounded Corners
* :toc: TableOfContents

* Variables
#+begin_src conf :tangle config
set $mod Mod4
set $term alacritty
set $menu rofi -show drun
set $GOPATH "/home/xanadul/go/bin"
set $wallpaper-path /home/xanadul/Pictures/Wallpaper/Falcom

set $lobin "~/.local/bin"
#+end_src

* Input and Output configuration
Since this general config should be used for 2 Systems (Laptop and Desktop), specific configuration gets tangled into different files, which can then be included as needed. Enables tap to click, 2-finger-tap rightclick, 3-finger-tap middle-click and correct 2-finger-scroll direction on the laptop touchpad.

Common config for keyboard Layout:
#+begin_src conf :tangle config
exec "setxkbmap de"
#+end_src

Output configuration for the desktop config.
#+begin_src conf :tangle config-desktop-specifics
#exec "~/.screenlayout/MainLayoutVega56"
#+end_src

Output and Touchpad config for the laptop config:
#+begin_src conf :tangle config-laptop-specifics
#+end_src

* Autostart
#+begin_src conf :tangle config
exec_always --no-startup-id "/home/xanadul/.config/i3/autostart.sh $wallpaper-path"
#Automount network filesystem over sshfs with keypair
exec "sshfs xanadul@mikoto.home:/mnt/Yukino/Data .Yukino"
#+end_src

* Keybinds
Important Keybinds:
| Keybind                | Action                                         | Keychord  | Action                       |
| Super+Return           | Runlauncher (set in variables, default rofi)   | Super+p   |                              |
| Super+t                | Terminal (set in variables), default alacritty | ->p       | Password manager (rofi-pass) |
| Super+c                | Closes focused container                       | ->i       | Password insert (rofi-pass)  |
| Super+Shift+r          | Reloads sway config                            | ->q       | Shutdown menu (XanLogout)    |
| Super+Shift+q          | Closes sway                                    | ->l       | Lock session                 |
| Super+(1-9)            | Move workspace to focused output               | ->h       | SuspendToRam                 |
| Super+Shift+(1-9)      | Move focused container to workspace            |           |                              |
| Super+(Arrows)         | Move the focus                                 | Super+s   |                              |
| Super+Shift+(Arrows)   | Move focused container in direction            | ->r       | Screenshot region            |
| Super+Control+(Arrows) | Resize container                               | ->w       | Screenshot window            |
|                        |                                                | ->a       | Screenshot all outputs       |
|                        |                                                | ->Shift+r | Record region                |
|                        |                                                | ->Shift+w | Record window                |
|                        |                                                | ->k       | Stop Recording               |
** i3 Management
#+begin_src conf :tangle config
#bindsym $mod+Shift+q exec "i3-msg exit"
bindsym $mod+Shift+q exit
bindsym $mod+Shift+r restart

#+end_src

** Applications

#+begin_src conf :tangle config
#Application launching
bindsym $mod+t exec $term
bindsym $mod+Return exec $menu
bindsym $mod+b exec "qutebrowser"
bindsym $mod+Shift+b exec "librewolf"
bindsym $mod+Shift+Return exec "dolphin"
bindsym $mod+e exec "emacs"

#Clipboard history
#bindsym Control+Shift+v exec "cliphist list | rofi -dmenu -i | cliphist decode | wtype -s 1 -"
#bindsym $mod+Control+v exec "$cliphist list | rofi -dmenu -i | cliphist delete"
#bindsym $mod+Control+Shift+v exec "cliphist wipe"

#Waybar toggle
#bindsym $mod+l exec "killall -SIGUSR1 waybar"

#+end_src
** Window-Management
#+begin_src conf :tangle config
#Stuff
bindsym $mod+c kill
floating_modifier $mod

#Focus
bindsym $mod+Left focus left
bindsym $mod+Down focus down
bindsym $mod+Up focus up
bindsym $mod+Right focus right

#Drawing tablet focus
bindsym $mod+Shift+Control+Left  exec "~/.local/bin/XanFocusTab left"
bindsym $mod+Shift+Control+Right exec "~/.local/bin/XanFocusTab right"

#Move Container
bindsym $mod+Shift+Left move left
bindsym $mod+Shift+Down move down
bindsym $mod+Shift+Up move up
bindsym $mod+Shift+Right move right

#Resize Container
bindsym $mod+Control+left  resize shrink   width   10px
bindsym $mod+Control+down  resize grow     height  10px
bindsym $mod+Control+up    resize shrink   height  10px
bindsym $mod+Control+right resize grow     width   10px

#Move Container to workspace
bindsym $mod+Shift+Control+1 exec "~/.local/bin/XanSwayMove 1"
bindsym $mod+Shift+Control+2 exec "~/.local/bin/XanSwayMove 2"
bindsym $mod+Shift+Control+3 exec "~/.local/bin/XanSwayMove 3"
bindsym $mod+Shift+Control+4 exec "~/.local/bin/XanSwayMove 4"
bindsym $mod+Shift+Control+5 exec "~/.local/bin/XanSwayMove 5"
bindsym $mod+Shift+Control+6 exec "~/.local/bin/XanSwayMove 6"
bindsym $mod+Shift+Control+7 exec "~/.local/bin/XanSwayMove 7"
bindsym $mod+Shift+Control+8 exec "~/.local/bin/XanSwayMove 8"
bindsym $mod+Shift+Control+9 exec "~/.local/bin/XanSwayMove 9"
bindsym $mod+Shift+Control+0 exec "~/.local/bin/XanSwayMove 0"

#Pull workspace to current output
bindsym $mod+1 exec "~/.local/bin/XanSwaySwitch 1 "
bindsym $mod+2 exec "~/.local/bin/XanSwaySwitch 2"
bindsym $mod+3 exec "~/.local/bin/XanSwaySwitch 3"
bindsym $mod+4 exec "~/.local/bin/XanSwaySwitch 4"
bindsym $mod+5 exec "~/.local/bin/XanSwaySwitch 5"
bindsym $mod+6 exec "~/.local/bin/XanSwaySwitch 6"
bindsym $mod+7 exec "~/.local/bin/XanSwaySwitch 7"
bindsym $mod+8 exec "~/.local/bin/XanSwaySwitch 8"
bindsym $mod+9 exec "~/.local/bin/XanSwaySwitch 9"
bindsym $mod+0 exec "~/.local/bin/XanSwaySwitch 0"

#Move Containe to workspace
bindsym $mod+Shift+1 move container to workspace number 1
bindsym $mod+Shift+2 move container to workspace number 2
bindsym $mod+Shift+3 move container to workspace number 3
bindsym $mod+Shift+4 move container to workspace number 4
bindsym $mod+Shift+5 move container to workspace number 5
bindsym $mod+Shift+6 move container to workspace number 6
bindsym $mod+Shift+7 move container to workspace number 7
bindsym $mod+Shift+8 move container to workspace number 8
bindsym $mod+Shift+9 move container to workspace number 9
bindsym $mod+Shift+0 move container to workspace number 10

#Layout commands
bindsym $mod+h splith
bindsym $mod+v splitv
bindsym $mod+Shift+w layout stacking
bindsym $mod+w layout tabbed
bindsym $mod+r layout toggle split
bindsym $mod+space focus mode_toggle
bindsym $mod+a focus parent
bindsym $mod+Shift+minus move scratchpad
bindsym $mod+minus scratchpad show
bindsym $mod+Shift+space floating toggle
bindsym $mod+f fullscreen

#Menus
mode "Menu" {
    # replace with rofi-pass: bindsym p mode "default" ; exec "tessen -a autotype"
    bindsym i mode "default" ; exec "rofi-pass --insert"
    bindsym q mode "default" ; exec "~/.local/bin/XanLogout"
    bindsym l mode "default" ; exec "i3lock -u -c 000000"
    bindsym h mode "default" ; exec "systemctl suspend"
    bindsym Escape mode "default"
}
mode "Screenshot" {
    bindsym r exec  --no-startup-id        '/home/xanadul/.local/bin/XanShoot' 'region'; mode "default"
    bindsym a exec   --no-startup-id       '~/.local/bin/XanShoot' 'all'; mode "default"
    bindsym w exec          '~/.local/bin/XanShoot' 'window'; mode "default"
    bindsym Shift+w exec  --no-startup-id  '/home/xanadul/.local/bin/XanShoot' 'window-video'; mode "default"
    bindsym Shift+r exec    '~/.local/bin/XanShoot' 'region-video'; mode "default"
    bindsym k mode "default"; exec '~/.local/bin/XanShoot' 'save-video'#to stop and save a recording, sending SIGINT to recorder
    bindsym Escape mode "default"
}
bindsym $mod+s mode "Screenshot"
bindsym $mod+p mode "Menu"
#+end_src

** Media and stuff
#+begin_src conf :tangle config
bindsym XF86AudioRaiseVolume            exec "pamixer -i 2"
bindsym XF86AudioLowerVolume            exec "pamixer -d 2"

bindsym Control+XF86AudioRaiseVolume    exec "~/.local/bin/XanMute sink"
bindsym Control+XF86AudioLowerVolume    exec "~/.local/bin/XanMute source"

bindsym XF86MonBrightnessUp             exec "xbacklight -inc 5"
bindsym XF86MonBrightnessDown           exec "xbacklight -dec 5"

#Switch between a mangohud config on the fly
bindsym $mod+M                          exec "~/.local/bin/XanSwitchMango"
#+end_src

* Settings
#+begin_src conf :tangle config
mouse_warping output
popup_during_fullscreen smart
workspace_auto_back_and_forth yes
#+end_src

* Theming
Window borders get managed by xborder, to achieve rounded corners.
#+begin_src conf :tangle config
title_align center
default_border pixel 4
gaps inner 10
#gaps outer 10
smart_gaps off
font pango:NotoMono 10

#Colors
#class                  border  backgr  text    indicator   child_border
client.focused          #bd93f9 #bd93f9 #282a36 #ffffff     #bd93f9
client.focused_inactive #6272a4 #6272a4 #f8f8f2 #ffffff     #6272a4
client.unfocused        #44475a #44475a #f8f8f2 #ffffff     #44475a
client.urgent           #ff5555 #ffff55 #282a36 #ffffff     #ffff55
#+end_src

* Bar
Basic bar. Will get replaced with better alternative like polybar.
#+begin_src conf :tangle config
        bar {
        status_command i3status
        position top
        tray_output none
        separator_symbol "|"
        colors {
            background #282a36
#                               border  bg      text
            focused_workspace   #a585f2 #bd93f9 #282a36
        }
}
#+end_src

* Includes
Useful for including the device specific config options. (Un)comment in tangled config as needed.
#+begin_src conf :tangle config

include ~/.config/i3/config-desktop-specifics
#include ~/.config/i3/config-laptop-specifics

#+end_src


dotconfig/nyxt/config.org
#+TITLE:Xanaduls Nyxt config
#+AUTHOR:Xanadul
#+STARTUP: showeverything
#+OPTIONS: toc:3
#+property: header-args :tangle config.lisp
#+auto_tangle: t

* :toc:
- [[#header][Header]]
- [[#nyxt-components][Nyxt Components]]
- [[#search-engines][Search Engines]]
- [[#hinting][Hinting]]
- [[#buffermodes-configs][Buffer/Modes Configs]]
  - [[#every-buffer][Every Buffer]]
  - [[#web-buffers][Web-Buffers]]
  - [[#browser][Browser]]
- [[#functions][Functions]]
  - [[#umpv][UMPV]]
- [[#hsplit][Hsplit]]
- [[#keybinds][Keybinds]]
  - [[#emacs][Emacs]]
  - [[#vi-normal][VI-Normal]]
- [[#search-engineslisp][search-engines.lisp]]

* Header
#+begin_src lisp 
(in-package #:nyxt-user)
(defvar *web-buffer-modes*
  '(:vi-normal-mode))
#+end_src

* Nyxt Components
#+begin_src lisp
(define-nyxt-user-system-and-load nyxt-user/basic-config
  :components ("keybinds" "hsplit"))
               ;;"passwd" "status" "commands" "style" "unpdf" "objdump" "github"))
#+end_src

* Search Engines
#+begin_src lisp
(defvar *my-search-engines*
  (list
   '("goog"     "https://google.com/search?q=~a" "https://google.com")
   '("python3"  "https://docs.python.org/3/search.html?q=~a" "https://docs.python.org/3")
   '("br"       "https://search.brave.com/search?q=~a" "https://search.brave.com")
   '("s"        "http://mikoto.ts:8380/search?q=~a&category_general=1&language=en-US&time_range=&safesearch=0&theme=simple" "http://mikoto.ts:3003")
   '("yt"       "https://www.youtube.com/results?search_query=~a")
   '("wiki"     "https://en.wikipedia.org/w/index.php?search=~a&title=Special%3ASearch&ns0=1"))
  "List of search engines.")
(define-configuration context-buffer
      "Go through the search engines above and make-search-engine out of them."
    ((search-engines
          (append
                (mapcar (lambda (engine) (apply 'make-search-engine engine))
                        *my-search-engines*)
                %slot-default%))))
#+end_src

* Hinting
#+begin_src lisp
(define-configuration :hint-mode
  "Set up Colmak-DH-Xan home row as the hint keys."
  ((hints-alphabet "TSNEGMPLFUWZ")))
#+end_src


* Buffer/Modes Configs
** Every Buffer
#+begin_src lisp
(define-configuration :buffer
    "Smooth scrolling in every buffer"
    ((smooth-scrolling t))
    ((default-modes "dark-mode")))
#+end_src

** Web-Buffers
#+begin_src lisp
(define-configuration :web-buffer
  "Basic modes setup for web-buffer."
  ((default-modes `(,@*web-buffer-modes* ,@%slot-value%))))
#+end_src

** Browser
#+begin_src lisp
(define-configuration :browser
 "Set new buffer URL (a.k.a. start page, new tab page)."
 ((default-new-buffer-url "https://search.brave.com/")))
#+end_src

* Functions
** UMPV
#+begin_src lisp
 ;; Create a function to launch mpv with given url
(defun mpv (url)
  "MPV launches with given url using the fast profile."
  (uiop:run-program (list "mpv" url)))
#+end_src

* Hsplit
#+begin_src lisp
(define-panel-command hsplit-internal (&key (url (quri:render-uri (url (current-buffer)))))
    (panel "*Duplicate panel*" :right)
  "Duplicate the current buffer URL in the panel buffer on the right.

A poor man's hsplit :)"
  (setf (ffi-width panel) 550)
  (run-thread "URL loader"
    (sleep 0.3)
    (buffer-load (quri:uri url) :buffer panel))
  "")

(define-command-global close-all-panels ()
  "Close all the panel buffers there are."
  (alexandria:when-let ((panels (nyxt/renderer/gtk::panel-buffers-right (current-window))))
    (delete-panel-buffer :window (current-window) :panels panels))
  (alexandria:when-let ((panels (nyxt/renderer/gtk::panel-buffers-left (current-window))))
    (delete-panel-buffer :window (current-window) :panels panels)))

(define-command-global hsplit ()
  "Based on `hsplit-internal' above."
  (if (nyxt/renderer/gtk::panel-buffers-right (current-window))
      (delete-all-panel-buffers (current-window))
      (hsplit-internal)))

#+end_src

* Keybinds
** Emacs
#+begin_src lisp
(define-configuration :document-mode
  "Add basic keybindings."
  ((keyscheme-map
    (keymaps:define-keyscheme-map
     "custom" (list :import %slot-value%)
     ;; If you want to have VI bindings overriden, just use
     ;; `scheme:vi-normal' or `scheme:vi-insert' instead of
     ;; `scheme:emacs'.
     nyxt/keyscheme:emacs
     (list "C-c p" 'copy-password
           "C-c y" 'autofill
           "C-f" :history-forwards-maybe-query
           "C-i" :input-edit-mode
           "M-:" 'eval-expression
           "C-s" :search-buffer
           "C-x 3" 'hsplit
           "C-x 1" 'close-all-panels
           "C-'"  (lambda-command insert-left-angle-quote ()
                    (ffi-buffer-paste (current-buffer) "«"))
           "C-M-'" (lambda-command insert-left-angle-quote ()
                     (ffi-buffer-paste (current-buffer) "»"))
           "C-M-hyphen" (lambda-command insert-left-angle-quote ()
                          (ffi-buffer-paste (current-buffer) "—"))
           "C-M-_" (lambda-command insert-left-angle-quote ()
                     (ffi-buffer-paste (current-buffer) "–"))
           "C-E" (lambda-command small-e-with-acute ()
                   (ffi-buffer-paste (current-buffer) "é"))
           "C-A" (lambda-command small-a-with-acute ()
                   (ffi-buffer-paste (current-buffer) "á"))
           "C-I" (lambda-command small-i-diaeresis ()
                   (ffi-buffer-paste (current-buffer) "ï"))
           "C-h hyphen" 'clcs-lookup)

#+end_src

** VI-Normal
#+begin_src lisp
     nyxt/keyscheme:vi-normal
     (list "space space" 'execute-command
	   "space b i" 'list-buffers
	   "space b b" 'switch-buffer
	   "space b C" 'delete-buffer
	   "space b c" 'delete-current-buffer
	   "space b u" 'reopen-last-buffer
	   "space b U" 'reopen-buffer
	   "space b h" 'list-history
	   
	   "space ." 'set-url
	   "space n" 'set-url-new-buffer

           "space t h" 'headings-panel
	   
	   "down" 'scroll-down
	   "shift-down" 'scroll-page-down
	   "up" 'scroll-up
	   "shift-up" 'scroll-page-up
 	   "f"          'follow-hint
 	   "F"          'follow-hint-new-buffer)))))

#+end_src

* search-engines.lisp

#+begin_src lisp :tangle no
in-package #:nyxt-user)

;; Define buffer search-engines slot to be a list of several
;; nx-search-engines-provided ones.
(define-configuration (buffer web-buffer)
  ((search-engines (list (engines:google :shortcut "gmaps"
                                         :object :maps)
                         (engines:wordnet :shortcut "wn"
                                          :show-word-frequencies t)
                         (engines:google :shortcut "g"
                                         :safe-search nil)
                         (engines:duckduckgo :theme :terminal
                                             :help-improve-duckduckgo nil
                                             :homepage-privacy-tips nil
                                             :privacy-newsletter nil
                                             :newsletter-reminders nil
                                             :install-reminders nil
                                             :install-duckduckgo nil)))))

#+end_src


dotconfig/nyxt/config.lisp
(in-package #:nyxt-user)
(defvar *web-buffer-modes*
  '(:vi-normal-mode))

(define-nyxt-user-system-and-load nyxt-user/basic-config
  :components ("keybinds" "hsplit"))
               ;;"passwd" "status" "commands" "style" "unpdf" "objdump" "github"))

(defvar *my-search-engines*
  (list
   '("goog"     "https://google.com/search?q=~a" "https://google.com")
   '("python3"  "https://docs.python.org/3/search.html?q=~a" "https://docs.python.org/3")
   '("br"       "https://search.brave.com/search?q=~a" "https://search.brave.com")
   '("s"        "http://mikoto.ts:8380/search?q=~a&category_general=1&language=en-US&time_range=&safesearch=0&theme=simple" "http://mikoto.ts:3003")
   '("yt"       "https://www.youtube.com/results?search_query=~a")
   '("wiki"     "https://en.wikipedia.org/w/index.php?search=~a&title=Special%3ASearch&ns0=1"))
  "List of search engines.")
(define-configuration context-buffer
      "Go through the search engines above and make-search-engine out of them."
    ((search-engines
          (append
                (mapcar (lambda (engine) (apply 'make-search-engine engine))
                        *my-search-engines*)
                %slot-default%))))

(define-configuration :hint-mode
  "Set up Colmak-DH-Xan home row as the hint keys."
  ((hints-alphabet "TSNEGMPLFUWZ")))

(define-configuration :buffer
    "Smooth scrolling in every buffer"
    ((smooth-scrolling t))
    ((default-modes "dark-mode")))

(define-configuration :web-buffer
  "Basic modes setup for web-buffer."
  ((default-modes `(,@*web-buffer-modes* ,@%slot-value%))))

(define-configuration :browser
 "Set new buffer URL (a.k.a. start page, new tab page)."
 ((default-new-buffer-url "https://search.brave.com/")))

;; Create a function to launch mpv with given url
(defun mpv (url)
  "MPV launches with given url using the fast profile."
  (uiop:run-program (list "mpv" url)))

(define-panel-command hsplit-internal (&key (url (quri:render-uri (url (current-buffer)))))
    (panel "*Duplicate panel*" :right)
  "Duplicate the current buffer URL in the panel buffer on the right.

A poor man's hsplit :)"
  (setf (ffi-width panel) 550)
  (run-thread "URL loader"
    (sleep 0.3)
    (buffer-load (quri:uri url) :buffer panel))
  "")

(define-command-global close-all-panels ()
  "Close all the panel buffers there are."
  (alexandria:when-let ((panels (nyxt/renderer/gtk::panel-buffers-right (current-window))))
    (delete-panel-buffer :window (current-window) :panels panels))
  (alexandria:when-let ((panels (nyxt/renderer/gtk::panel-buffers-left (current-window))))
    (delete-panel-buffer :window (current-window) :panels panels)))

(define-command-global hsplit ()
  "Based on `hsplit-internal' above."
  (if (nyxt/renderer/gtk::panel-buffers-right (current-window))
      (delete-all-panel-buffers (current-window))
      (hsplit-internal)))

(define-configuration :document-mode
  "Add basic keybindings."
  ((keyscheme-map
    (keymaps:define-keyscheme-map
     "custom" (list :import %slot-value%)
     ;; If you want to have VI bindings overriden, just use
     ;; `scheme:vi-normal' or `scheme:vi-insert' instead of
     ;; `scheme:emacs'.
     nyxt/keyscheme:emacs
     (list "C-c p" 'copy-password
           "C-c y" 'autofill
           "C-f" :history-forwards-maybe-query
           "C-i" :input-edit-mode
           "M-:" 'eval-expression
           "C-s" :search-buffer
           "C-x 3" 'hsplit
           "C-x 1" 'close-all-panels
           "C-'"  (lambda-command insert-left-angle-quote ()
                    (ffi-buffer-paste (current-buffer) "«"))
           "C-M-'" (lambda-command insert-left-angle-quote ()
                     (ffi-buffer-paste (current-buffer) "»"))
           "C-M-hyphen" (lambda-command insert-left-angle-quote ()
                          (ffi-buffer-paste (current-buffer) "—"))
           "C-M-_" (lambda-command insert-left-angle-quote ()
                     (ffi-buffer-paste (current-buffer) "–"))
           "C-E" (lambda-command small-e-with-acute ()
                   (ffi-buffer-paste (current-buffer) "é"))
           "C-A" (lambda-command small-a-with-acute ()
                   (ffi-buffer-paste (current-buffer) "á"))
           "C-I" (lambda-command small-i-diaeresis ()
                   (ffi-buffer-paste (current-buffer) "ï"))
           "C-h hyphen" 'clcs-lookup)

nyxt/keyscheme:vi-normal
(list "space space" 'execute-command
      "space b i" 'list-buffers
      "space b b" 'switch-buffer
      "space b C" 'delete-buffer
      "space b c" 'delete-current-buffer
      "space b u" 'reopen-last-buffer
      "space b U" 'reopen-buffer
      "space b h" 'list-history

      "space ." 'set-url
      "space n" 'set-url-new-buffer

      "space t h" 'headings-panel

      "down" 'scroll-down
      "shift-down" 'scroll-page-down
      "up" 'scroll-up
      "shift-up" 'scroll-page-up
      "f"          'follow-hint
      "F"          'follow-hint-new-buffer)))))
